storage-node/src/services/sync/acceptPendingObjects.ts

```diff
@@ -0,0 +1,212 @@
+
+    await this.loadPendingDataObjectsFromIDs(pendingIds)
+
+    logger.debug(`Pending data objects ID cache loaded.`)
+  }
+
+  public push(dataObjectId: string, storageBucketId: string, bagId: string): void {
+    this.pendingDataObjects.set(dataObjectId, [storageBucketId, bagId])
+  }
+
+  private async popN(n: number): Promise<AcceptPendingDataObjectsParams[]> {
+    const params: AcceptPendingDataObjectsParams[] = []
+    let count = 0
+
+    while (count < n && this.pendingDataObjects.size > 0) {
+      // Extract the first element from the map
+      const [dataObjectId, [storageBucketId, bagId]]: [string, [string, string]] = this.pendingDataObjects
+        .entries()
+        .next().value
+      this.pendingDataObjects.delete(dataObjectId)
+
+      // Find or create the storage bucket in the params array
+      let storageBucket = params.find((p) => p.storageBucket.id === storageBucketId)
+      if (!storageBucket) {
+        const account = this.bucketKeyPairs.get(storageBucketId.toString())
+        if (!account) {
+          logger.error(`No key pair found for storage bucket ${storageBucketId}.`)
+          continue
+        }
+
+        storageBucket = {
+          account,
+          storageBucket: {
+            id: storageBucketId,
+            bags: [],
+          },
+        }
+        params.push(storageBucket)
+      }
+
+      // Find or create the bag in the storage bucket in the params array
+      let bag = storageBucket.storageBucket.bags.find((b) => b.id === bagId)
+      if (!bag) {
+        bag = {
+          id: bagId,
+          dataObjects: [],
+        }
+        storageBucket.storageBucket.bags.push(bag)
+      }
+
+      // Add the data object to the bag in the params array
+      bag.dataObjects.push(dataObjectId)
+
+      count++
+    }
+
+    return params
+  }
+
+  /**
+   * Returns file names from the pending objects directory.
+   *
+   * @param directory - local directory to get file names from
+   */
+  private async getPendingObjectsFromLocalDir(directory: string): Promise<string[]> {
+    // Check if directory exists and if not, create it
+    if (!fs.existsSync(directory)) {
+      await fsPromises.mkdir(directory)
+    }
+
+    // Read the directory contents
+    return await fsPromises.readdir(directory)
+  }
+
+  private async acceptPendingDataObjects(
+    api: ApiPromise,
+    workerId: number,
+    uploadsDir: string,
+    maxTxBatchSize: number
+  ): Promise<void> {
+    const params = await this.popN(maxTxBatchSize)
+
+    let failedObjectsIds: string[]
+    try {
+      failedObjectsIds = await acceptPendingDataObjectsBatch(api, workerId, params)
+    } catch (err) {
+      // Re-push all params as failed since the entire batch failed
+      params.forEach((param) => {
+        param.storageBucket.bags.forEach((bag) => {
+          bag.dataObjects.forEach((dataObjectId) => {
+            this.push(dataObjectId, param.storageBucket.id, bag.id)
```

Comment:
Is it possible the bag doesn't belong in the bucket anymore, and we should let the service drop it since the transaction will never work?

Reply:
Do you mean runtime transaction would fail if any of the calls are invalid?  If yes, that will not be the case, since we are using [forceBatch](https://github.com/zeeshanakram3/joystream/blob/51901e12660b6e7cd5589cd29443da7a6db17bd5/storage-node/src/services/runtime/extrinsics.ts#L167)

Reply:
Yes, in the sense that it will remain in pending folder and will be retried again later. But I think this is similar to discussion here https://github.com/Joystream/joystream/pull/4971#discussion_r1409373556